### PR TITLE
Update pre-commit to include git hook restage and .only file checks

### LINKED ISSUE #19: Add linting rules to check test content
There are some additional linting checks that should be included:

- [ ] fix(dev): check no `.only` in test files
  - Linting during the `pre-commit` git hook should check any `.cy.js` files being committed to ensure that they do not contain a `.only` on a test.  The `pre-commit` hook should fail if any `cy.js` file contains a `.only`.
- [ ] fix(dev): check no `console.log` in test files
  - Linting during the `pre-commit` git hook should check any `.cy.js` files being committed to ensure that they do not contain `console.log` statements.  The `pre-commit` hook should fail if any `cy.js` file contains a `console.log` statement.

### LINKED ISSUE #32: Have `pre-commit` git hook re-add staged files
When the `pre-commit` git hook fails and you go make a correction the changed files need to be staged again.  it is easy to forget to do that extra `git stage`.  The `pre-commit` hook should check if any of the staged files have been modified and then ask if they should be staged again.

### LINKED ISSUE #33: Improve efficiency of `lintstaged` that is run in git pre-commit hook
Currently if a `.vue` file (i.e. a component or entry point) changes all of its tests are also run.  If any of the tests for the component have also changed then they are run a second time.  So, if a `.vue` file is changed, we can skip running any of the changed tests because they will have already been run.  

A similar inefficiency occurs with changes to any of the library files (in `library` or in `lib.js` within the modules).  These inefficiencies should be addressed as well.

### PR DESCRIPTION
**Pull request by WarpWing**
**Pull Request Description**

This PR aims to fix #19 , #32 and #33 by implementing checks for `.only` files, readding staged files and reduce redundant `.vue` test runs.

Closes #19
Closes #32
Closes #33 

---

**Licensing Certification**

FarmData2 is a [Free Cultural Work](https://freedomdefined.org/Definition) and all accepted contributions are licensed as described in the LICENSE.md file. This requires that the contributor holds the rights to do so. By submitting this pull request **I certify that I satisfy the terms of the [Developer Certificate of Origin](https://developercertificate.org/)** for its contents.

### CODE REVIEW

**Comment by braughtg** [comment]
I'll have a look at this ~~tomorrow~~ soon.  In the meantime, you should be sure to run it through the linters, etc, if you hadn't already.  Two ways to do that:
- manually: `npm run lint`
  - Will lint/format check the entire project.
- via git-hook: 
  - change into `.git`
  - delete the `hooks` directory, 
  - make a symbolic link named `hooks` to `../.githooks`
  - then the files in any commit will be linted/styled/tested before being committed.

File: .githooks/pre-commit [hunk 1]
```diff
@@ -2,28 +2,53 @@
...
-  echo
-  exit 255
-fi
-
-if [ "$BRANCH" = "development" ]; then
-  echo
-  echo "You cannot commit directly to the development branch."
-  echo "Instead create a feature branch and commit your changes to it."
+  echo "You cannot commit directly to the $BRANCH branch."
+  echo "Instead, create a feature branch and commit your changes to it."
```
Comment (braughtg) [inline .githooks/pre-commit:10]: "Instead, create a feature branch from the development branch and commit your changes to it."

File: .githooks/pre-commit [hunk 2]
```diff
@@ -2,28 +2,53 @@
...
+      skipTests+=("$file")
+    fi
+  fi
+done
+
+# Check for any modifications in already staged files
+for file in $(git diff --cached --name-only); do
+  if git diff --name-only | grep -q "$file"; then
+    echo "File $file is staged but has uncommitted changes."
+    read -p "Would you like to restage it? [Y/n]: " -n 1 -r
```
Comment (braughtg) [inline .githooks/pre-commit:41]: "Would you like to restage it before committing?"

**Comment by braughtg** [comment]
To review:

- Pull the PR.

For #19:
- [ ] - Modify a test file to add `.only` to one test, stage and commit.  Commit should be rejected because of `.only` in the test file.
- [ ] - Modify several test files to contain `.only` tests.  Ensure at least one file has multiple `.only`. Stage and commit. Ensure that commit is rejected until all `.only` are removed.

For #32:
- [ ] - Make change to a file will cause the commit to be rejected.  
- [ ] - Fix the file and save, but don't stage, and attempt commit again.
- [ ] - Should be prompted to restage file.
- [ ] - Respond no to restage and ensure that commit fails again.
- [ ] - Respond yes and ensure that commit is accepted.
- [ ] - Repeat with multiple files.  
  - [ ] - Try restaging some but not all and ensure commit fails.
  - [ ] - Try restaging all and ensure that commit passes.

For #33:
- [ ] - Modify a `.vue` file. Stage and commit.  Ensure all  `.cy.js` tests in directory with the `.vue` still run.
- [ ] - Modify a `.cy.js` file associated with a `.vue` file. Stage and commit. Ensure that the changed `cy.js` tests run.
- [ ] - Modify a `.vue` file and an associated `.cy.js` file.  Stage and commit. 
  - [ ] - Ensure all  `.cy.js` tests in directory with the `.vue` still run.
  - [ ] - Ensure that the modified `cy.js` files are not run a second time.
- [ ] - Modify multiple `.vue` and multiple `.cy.js` associated with each.
  - [ ] - Ensure all `.cy.js` tests in the directory with the `.vue` files run.
  - [ ] - Ensure that the modified `cy.js` files are not run a second time.

File: .githooks/pre-commit [hunk 3]
```diff
@@ -2,28 +2,53 @@
...
   echo
   exit 255
 fi
 
-# For all other branches, run the pre-commit checks using lint-staged.
-if npm run check:staged; then
+# Initialize an array to keep track of tests that do not need to be run again.
+declare -a skipTests=()
+
+# Check for .only in .cy.js files and mark .vue file tests
```
Comment (braughtg) [inline .githooks/pre-commit:19]: See my comment in the overall review comment about possibly using an eslint rule for this instead.

**Review by braughtg — CHANGES_REQUESTED** [review]
The approaches to restaging of files and to not rerunning tests look good (though I have not fully tested them yet).  

For preventing `.only` tests from being commited, I think it would be preferable to use an eslint rule rather than custom scripting.  By adding the lint rule devs will be able to see the warning in VSCode and when running linting before seeing it in the pre-commit hook.  Also in the pre-commit hook the same check will be run just as a regular part of the linting done by `check:staged` - so no modifications to the pre-commit hook will be needed for this part once the rule is added.

Maybe give this plugin a try and see how it does:
- eslint-plugin-no-only-tests: https://www.npmjs.com/package/eslint-plugin-no-only-tests

Once we have that in place, I'll be using the steps above for review.  So might be worth trying them in advance.

Flip this back to a full-PR (rather than a draft) when its ready for review.  Thanks!

**Comment by braughtg** [comment]
My PR #79 made some changes to the `test.bash` and `lintstagedrc.cjs` file that may need to be considered in continued work on this PR.  In particular:
- end-to-end tests all now have the extension `*.e2e.cy.js`
- unit tests are now supported for `lib.js` with tests named `lib.*.unit.cy.js` in module endpoints.  So these have the same issue with re-running test as the others described in #33.

**Comment by WarpWing** [comment]
I'll circle back to this but since there are some merge conflicts, I'm going to delete this and patch my changes back.

**Pull request closed without merging**